Block vectors and the update_ghost_values function

I noticed this while working on #4297. Consider the following template function in `block_vector_base.h`:

```cpp
2087 template <class VectorType>
2088 void BlockVectorBase<VectorType>::update_ghost_values () const
2089 {
2090   for (size_type i=0; i<n_blocks(); ++i)
2091     block(i).update_ghost_values ();
2092 }
```

`PETScWrappers::MPI::Vector` does not implement this function, so trying to compile

```cpp
#include <deal.II/lac/petsc_parallel_block_vector.h>

int main()
{
  using namespace dealii;

  PETScWrappers::MPI::BlockVector bv;
  bv.update_ghost_values();
}
```

results in an error:

```
[drwells@archway petsc-wrapper-test]$ make
[ 50%] Building CXX object CMakeFiles/petsct.dir/petsct.cc.o
In file included from [...]/include/deal.II/lac/petsc_parallel_block_vector.h:26:0,
                 from [...]/petsc-wrapper-test/petsct.cc:1:
[...]/include/deal.II/lac/block_vector_base.h: In instantiation of
'void dealii::BlockVectorBase< <template-parameter-1-1> >::update_ghost_values() const
[with VectorType = dealii::PETScWrappers::MPI::Vector]':
/home/drwells/Documents/Code/CPP/petsc-wrapper-test/petsct.cc:11:26:   required from here
/home/drwells/Applications/dealii-dev/include/deal.II/lac/block_vector_base.h:2083:14:
error: 'const BlockType {aka const class dealii::PETScWrappers::MPI::Vector}'
has no member named 'update_ghost_values'
     block(i).update_ghost_values ();
     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/petsct.dir/build.make:63: CMakeFiles/petsct.dir/petsct.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:196: CMakeFiles/petsct.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

However, our own linear algebra classes implement this function (and so do our Trilinos wrappers). We should either implement a stub or re-implement this function for the PETSc vector.

--- a/include/deal.II/lac/petsc_vector_base.h
+++ b/include/deal.II/lac/petsc_vector_base.h
@@ -370,6 +370,14 @@ namespace PETScWrappers
     bool has_ghost_elements() const;
 
     /**
+     * This function only exists for compatibility with the @p
+     * LinearAlgebra::distributed::Vector class and does nothing: this class
+     * implements ghost value updates in a different way that is a better fit
+     * with the underlying PETSc vector object.
+     */
+    void update_ghost_values () const;
+
+    /**
      * Provide access to a given element, both read and write.
      */
     reference
@@ -1104,6 +1112,13 @@ namespace PETScWrappers
   {
     return ghosted;
   }
+
+
+
+  inline
+  void
+  VectorBase::update_ghost_values () const
+  {}
 
 
 

--- a/include/deal.II/lac/trilinos_vector_base.h
+++ b/include/deal.II/lac/trilinos_vector_base.h
@@ -413,6 +413,14 @@ namespace TrilinosWrappers
     bool has_ghost_elements() const;
 
     /**
+     * This function only exists for compatibility with the @p
+     * LinearAlgebra::distributed::Vector class and does nothing: this class
+     * implements ghost value updates in a different way that is a better fit
+     * with the underlying Trilinos vector object.
+     */
+    void update_ghost_values () const;
+
+    /**
      * Return the scalar (inner) product of two vectors. The vectors must have
      * the same size.
      */
@@ -1015,6 +1023,13 @@ namespace TrilinosWrappers
   {
     return has_ghosts;
   }
+
+
+
+  inline
+  void
+  VectorBase::update_ghost_values () const
+  {}
 
 
 

